### PR TITLE
Add inbounds annotation in `map!` computation

### PR DESCRIPTION
**Pull request by YingboMa**
```julia
using BenchmarkTools

function foo!(a, b, c, d, e)
    map!((b, c, d, e) -> b + 0.1 * (0.2c + 0.3d + 0.4e), a, b, c, d, e)
end

a,b,c,d,e=(ones(1000) for i in 1:5)
@btime foo!($a, $b, $c, $d, $e);
# Master: 1.753 μs (1 allocation: 48 bytes)
# PR: 347.575 ns (1 allocation: 48 bytes)
```

If I put `@_propagate_inbounds_meta` in `map_n!`, the `map!` function will be a lot slower

```julia
julia> @btime foo!($a, $b, $c, $d, $e);
  583.367 ns (1 allocation: 48 bytes)
```
.

### CODE REVIEW

**Comment by nalimilan** [comment]
Thanks. I'd say we should merge the PR without the `FIXME` part for now, we can always figure out later why adding the `@_propagate_inbounds_meta` makes things (slightly) slower. As a first contribution, `map_n!` was introduced by @JeffBezanson in f4ed999cb73293878b2e9b98fee415548533cd5a, making it a separate function from `map!`, probably because it was faster for a reason that escapes me. So it's not completely surprising that inlining it makes things slower.

File: base/abstractarray.jl [hunk 1]
```diff
@@ -2057,10 +2057,17 @@ end
...
+    @_propagate_inbounds_meta
+    return (as[1][i], ith_all(i, tail(as))...)
+end
 
 function map_n!(f::F, dest::AbstractArray, As) where F
-    for i = LinearIndices(As[1])
+    #FIXME: adding `@_propagate_inbounds_meta` will destroy the performance of `map!`
+    #@_propagate_inbounds_meta
+    idxs1 = LinearIndices(As[1])
+    @boundscheck LinearIndices(dest) == idxs1 && all(x -> x == idxs1, map(LinearIndices, As))
```
Comment (nalimilan) [inline base/abstractarray.jl]: Could be written 
```suggestion
    @boundscheck LinearIndices(dest) == idxs1 && all(x -> LinearIndices(x) == idxs1, tail(As))
```

File: base/abstractarray.jl [hunk 2]
```diff
@@ -2009,7 +2009,7 @@ concatenate_setindex!(R, X::AbstractArray, I...) = (R[I...] = X)
 
 function map!(f::F, dest::AbstractArray, A::AbstractArray) where F
     for (i,j) in zip(eachindex(dest),eachindex(A))
-        dest[i] = f(A[j])
+        @inbounds dest[i] = f(A[j])
```
Comment (vtjnash) [inline base/abstractarray.jl]: I don't think `f()` is inbounds-guaranteed here. For example, if it was `i -> [][i]`, that would be always OOB.
```suggestion
        val = f(@inbounds A[j])
        @inbounds dest[i] = val
```

File: base/abstractarray.jl [hunk 3]
```diff
@@ -2057,10 +2057,17 @@ end
...
+    return (as[1][i], ith_all(i, tail(as))...)
+end
 
 function map_n!(f::F, dest::AbstractArray, As) where F
-    for i = LinearIndices(As[1])
+    #FIXME: adding `@_propagate_inbounds_meta` will destroy the performance of `map!`
+    #@_propagate_inbounds_meta
+    idxs1 = LinearIndices(As[1])
+    @boundscheck LinearIndices(dest) == idxs1 && all(x -> x == idxs1, map(LinearIndices, As))
+    @inbounds for i = idxs1
```
Comment (vtjnash) [inline base/abstractarray.jl]: same here

File: base/abstractarray.jl (hunk 1 above)
Comment (YingboMa) [reply base/abstractarray.jl]:
```julia
julia> @btime foo!($a, $b, $c, $d, $e); # with tail
  371.754 ns (2 allocations: 80 bytes)

julia> @btime foo!($a, $b, $c, $d, $e); # without tail
  366.971 ns (1 allocation: 48 bytes)
```
`tail` introduces a little more allocation, I will use
```julia
all(x -> LinearIndices(x) == idxs1, As)
```
.

**Review by nalimilan — APPROVED** [review]

File: base/abstractarray.jl [hunk 4]
```diff
@@ -2049,19 +2050,28 @@ map(f, ::AbstractSet) = error("map is not defined on sets")
...
+        @inbounds dest[i] = val
     end
     return dest
 end
 
 ## N argument
 
 @inline ith_all(i, ::Tuple{}) = ()
-@inline ith_all(i, as) = (as[1][i], ith_all(i, tail(as))...)
+function ith_all(i, as)
```
Comment (JeffBezanson) [inline base/abstractarray.jl]: Are we sure it's ok to drop `@inline` here?

Comment (nalimilan) [reply base/abstractarray.jl]: Doesn't `@_propagate_inbounds_meta` imply inlining?

Comment (YingboMa) [reply base/abstractarray.jl]: https://github.com/JuliaLang/julia/blob/7772486a6942c9a9b52849b6ffe690a7da5861cd/base/essentials.jl#L99-L101

**Comment by YingboMa** [comment]
Bump

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/30624#issuecomment-455819095) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/61638463fb4004f3f15fdeb53c1232fbaffd968c/4a9fcf5_vs_b8c0ec8/report.md). cc @ararslan

**Comment by YingboMa** [comment]
The regression seems to be noise.

**Pull request merged**